### DISCUSSION TITLE
Arabic, Chinese, Thai, Persion, Japanese Langugaes not supporting

### DISCUSSION THREAD

**Discussion by arjun-dasinfomedia** [dompdf/dompdf#2933] [Q&A]
Arabic, Chinese, Thai, Persian, and Japanese Languages are not supported. It is showing ????? and square boxes, And in Arabic language showing wrong words.
[QUOTATION#J000037 (13).pdf](https://github.com/dompdf/dompdf/files/8967413/QUOTATION.J000037.13.pdf)
[QUOTATION#J000037 (14).pdf](https://github.com/dompdf/dompdf/files/8967416/QUOTATION.J000037.14.pdf)

**Comment by bsweeney** [comment]
First, you have to install a font (or fonts) that supports your target languages. Read through the [Unicode How-To](https://github.com/dompdf/dompdf/wiki) for more information.

Second, Dompdf does not currently support complex text layout operations (bidi, shaping, ordering). The appearance of words can be fixed up prior to PDF generation by pre-process with a library that supports the language ([tc-lib-unicode](https://github.com/tecnickcom/tc-lib-unicode), [Ar-PHP](https://github.com/khaled-alshamaa/ar-php)). Some relevant discussion for Arabic can be found in issue #712.

Follow #2619 for updates on native support for this type of functionality.
